core/src/impl/Kokkos_Combined_Reducer.hpp

```diff
@@ -0,0 +1,664 @@
+      const noexcept {
+    emulate_fold_comma_operator(
+        (reducers_that_reference_original_values.view()() =
+             m_value.template get<Idxs, typename Reducers::value_type>())...);
+  }
+};
+
+// Apparently this can't be an alias template because of a bug/unimplemented
+// feature in GCC's name mangler.  But in this case, this amounts to the same
+// thing.
+template <class Space, class... Reducers>
+struct CombinedReducer
+    : CombinedReducerImpl<make_index_sequence<sizeof...(Reducers)>, Space,
+                          Reducers...> {
+  using base_t = CombinedReducerImpl<make_index_sequence<sizeof...(Reducers)>,
+                                     Space, Reducers...>;
+  using base_t::base_t;
+  using reducer = CombinedReducer<Space, Reducers...>;
+};
+
+// </editor-fold> end CombinedReducer }}}1
+//==============================================================================
+
+//==============================================================================
+// <editor-fold desc="CombinedReductionFunctorWrapper"> {{{1
+
+template <class IdxSeq, class Functor, class Space, class... Reducers>
+struct CombinedReductionFunctorWrapperImpl;
+
+template <size_t... Idxs, class Functor, class Space, class... Reducers>
+struct CombinedReductionFunctorWrapperImpl<integer_sequence<size_t, Idxs...>,
+                                           Functor, Space, Reducers...> {
+ private:
+  Functor m_functor;
+
+ public:
+  //------------------------------------------------------------------------------
+  // <editor-fold desc="type aliases"> {{{2
+
+  using reducer_type = CombinedReducer<Space, Reducers...>;
+
+  // Prevent Kokkos from attempting to deduce value_type
+  using value_type = typename reducer_type::value_type;
+
+  // </editor-fold> end type aliases }}}2
+  //------------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="Ctors, destructor, and assignment"> {{{2
+
+  KOKKOS_INLINE_FUNCTION
+  constexpr CombinedReductionFunctorWrapperImpl() noexcept = default;
+  KOKKOS_INLINE_FUNCTION
+  constexpr CombinedReductionFunctorWrapperImpl(
+      CombinedReductionFunctorWrapperImpl const&) = default;
+  KOKKOS_INLINE_FUNCTION
+  constexpr CombinedReductionFunctorWrapperImpl(
+      CombinedReductionFunctorWrapperImpl&&) = default;
+  KOKKOS_INLINE_FUNCTION
+  KOKKOS_CONSTEXPR_14 CombinedReductionFunctorWrapperImpl& operator=(
+      CombinedReductionFunctorWrapperImpl const&) = default;
+  KOKKOS_INLINE_FUNCTION
+  KOKKOS_CONSTEXPR_14 CombinedReductionFunctorWrapperImpl& operator=(
+      CombinedReductionFunctorWrapperImpl&&) = default;
+  KOKKOS_INLINE_FUNCTION
+  ~CombinedReductionFunctorWrapperImpl() = default;
+
+  KOKKOS_INLINE_FUNCTION
+  constexpr explicit CombinedReductionFunctorWrapperImpl(Functor arg_functor)
+      : m_functor(std::move(arg_functor)) {}
```

Comment:
Honestly when I see all that noise with defaulted special members and other constructors, I am tempted to say "why don't you just make it a `struct` with all data member public and rely on aggregate initialization to construct?"

Reply:
Because aggregate initialization doesn't work consistently across all of the compilers we currently have to support.